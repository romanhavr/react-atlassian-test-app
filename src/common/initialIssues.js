export const initialIssues = [
  {
    id: 1,
    content: {
      id: 1,
      issue: '1-st',
      priority: { level: 2, label: 'Medium', value: 'medium' },
      assignee: 1,
      labelIds: [1]
    }
  },
  {
    id: 2,
    content: {
      id: 2,
      issue: '2-nd',
      priority: { level: 3, label: 'Low', value: 'low' },
      assignee: 2,
      labelIds: [2]
    }
  },
  {
    id: 3,
    content: {
      id: 3,
      issue: '1-st copy',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 1,
      labelIds: [1]
    }
  },
  {
    id: 4,
    content: {
      id: 4,
      issue: '3-rd',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 4,
      labelIds: [2, 3]
    }
  },
  {
    id: 5,
    content: {
      id: 5,
      issue: '1-st another copy',
      priority: { level: 3, label: 'Low', value: 'low' },
      assignee: 3,
      labelIds: [1, 3]
    }
  },
  {
    id: 6,
    content: {
      id: 6,
      issue: '1-st',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 4,
      labelIds: [1, 3]
    }
  },
  {
    id: 7,
    content: {
      id: 7,
      issue: '1-st',
      priority: { level: 2, label: 'Medium', value: 'medium' },
      assignee: 2,
      labelIds: [1, 3]
    }
  }
]