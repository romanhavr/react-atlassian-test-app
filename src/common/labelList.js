export const labels = [
    { id: 1, name: 'label1', label: 'Label 1' },
    { id: 2, name: 'anotherLabel', label: 'Another Label' },
    { id: 3, name: 'newLabel', label: 'New Label' },
]