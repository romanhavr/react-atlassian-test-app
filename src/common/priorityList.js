export const priorityList = [
    { level: 1, label: 'High', value: 'high' },
    { level: 2, label: 'Medium', value: 'medium' },
    { level: 3, label: 'Low', value: 'low' }
]