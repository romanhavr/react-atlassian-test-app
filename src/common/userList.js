// @flow

import type { User } from '../interfaces/interfaces';

export const userList: User[] = [
    {
        id: 1,
        displayName: 'Admin',
        avatar: 'https://pickaface.net/gallery/avatar/unr_edwinavator_170412_0152_yoidv.png'
    },
    {
        id: 2,
        displayName: 'Elvis',
        avatar: 'https://pickaface.net/gallery/avatar/Nirvana53183d8f601d7.png'
    },
    {
        id: 3,
        displayName: 'Roling Stone',
        avatar: 'https://tradiesinbusiness.com.au/wp-content/uploads/2019/09/avatar-jongen-voorbeeld-1.jpg'
    },
    {
        id: 4,
        displayName: 'John Smith',
        avatar: 'https://apollodesign.nl/wp-content/uploads/2015/08/avatar-casper-tekening.png'
    },
]