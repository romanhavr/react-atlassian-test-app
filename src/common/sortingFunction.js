// @flow

import type { TableItem, SortingOptions, SortingCriteria } from '../interfaces/interfaces';
import { priorityList } from '../common/priorityList';
import { userList } from '../common/userList';

export function sortingFunc(a: TableItem, b: TableItem, sortingBy: SortingOptions, i: number): number {
        if (comparing(a, b, sortingBy[i].value, 'more')) {
            return 1
        } else if (comparing(a, b, sortingBy[i].value, 'less')) {
            return -1
        } else if (comparing(a, b, sortingBy[i].value, 'eq') && i < sortingBy.length - 1) {
            return sortingFunc(a, b, sortingBy, ++i)
        } else
            return 0
};

function comparing (a: TableItem, b: TableItem, sortingValue: string, compareIndex: string): boolean {
    const newObject: SortingCriteria = sortingCriteria (a.content[sortingValue], b.content[sortingValue], sortingValue)
    switch (compareIndex) {
        case 'more' :
            return newObject.newObj1 > newObject.newObj2;
        case 'eq' :
            return newObject.newObj1 === newObject.newObj2;
        case 'less' :
            return newObject.newObj1 < newObject.newObj2;
        default:
            return newObject.newObj1 > newObject.newObj2;
    }
}

function sortingCriteria (obj1, obj2, sortingValue: string): SortingCriteria {
    switch(sortingValue) {
        case 'issue' :
            return { newObj1: obj1, newObj2: obj2};
        case 'assignee' :
            const userById1 = userList.find(el => el.id === obj1);
            const userById2 = userList.find(el => el.id === obj2);
            if (userById1 && userById2) return { 
                newObj1: userById1.displayName,
                newObj2: userById2.displayName
            };
            break;
        case 'priority' :
            const priorityByValue1 = priorityList.find(el => el.value === obj1.value);
            const priorityByValue2 = priorityList.find(el => el.value === obj2.value);
            if (priorityByValue1 && priorityByValue2) return {
                newObj1: priorityByValue1.level,
                newObj2: priorityByValue2.level
            };
            break;
        default:
            return { newObj1: obj1, newObj2: obj2};
    }
}