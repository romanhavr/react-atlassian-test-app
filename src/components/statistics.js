// @flow

import React from 'react';
import { connect } from 'react-redux';
import * as R from 'ramda';
import TableTree from '@atlaskit/table-tree';
import { getUniqueIssues, getUniqueIssuesBy, getUniqueLabelsBy, 
        getUniqueAssigneeBy, getUniquePriorityBy
} from '../store/selectors';
import { userList } from '../common/userList';
import { labels } from '../common/labelList';
import { priorityList } from '../common/priorityList';

type Props = {
    uniqueIssues: any,
    uniqueIssuesBy: any,
    uniqueAssigneeBy: any,
    uniqueLabelsBy: any,
    uniquePriorityBy: any
}

export function Statistics(props: Props) {
    const numberOf = (item) => <span>{item[0]}</span>;

    const numberBy = (item) => Object.entries(detailsBy(item[1])).map(entry => 
                <span key={entry[0]}>
                    <b>{entry[0]}:</b> {R.toString(entry[1])}<br/>
                </span>)
    
    const total = (item) => <span>{item[2]}</span>

    const stitisticsTableData = [
        {
            id: 'issues',
            content: ['Unique issues', 'issue', props.uniqueIssues.length]
        },
        {
            id: 'assegnees',
            content: ['Unique assignees', 'assignee', userList.length]
        },
        {
            id: 'labels',
            content: ['Unique labels', 'labelIds', labels.length]
        },
        {
            id: 'priority',
            content: ['Unique priorities', 'priority', priorityList.length]
        }
    ];

    const detailsBy = (key) => {
        switch(key) {
            case 'issue':
                return props.uniqueIssuesBy;
            case 'labelIds':
                return props.uniqueLabelsBy;
            case 'assignee':
                return props.uniqueAssigneeBy;
            case 'priority':
                return props.uniquePriorityBy;
            default:
                return props.uniqueIssues
        }
    }

    return (
        <React.Fragment>
            <h2>Statistics</h2>
            {/* {getUnique} */}
            <TableTree
                headers={['Number of', 'Number by', 'Total number']}
                columns={[numberOf, numberBy, total]}
                columnWidths={['300px', '300px', '200px']}
                items={stitisticsTableData}
            />
        </React.Fragment>
    );
}
const mapStateToProps = (state) => {
    return {
        uniqueIssues: getUniqueIssues(state),
        uniqueIssuesBy: getUniqueIssuesBy(state),
        uniqueAssigneeBy: getUniqueAssigneeBy(state),
        uniqueLabelsBy: getUniqueLabelsBy(state),
        uniquePriorityBy: getUniquePriorityBy(state),
    }
}

export default connect(
    mapStateToProps
)(Statistics);
