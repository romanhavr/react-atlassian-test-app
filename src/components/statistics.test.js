// @flow

import { shallow } from 'enzyme';
import React from 'react';
import { Statistics } from './statistics';
import { userList } from '../common/userList';
import { labels } from '../common/labelList';
import { priorityList } from '../common/priorityList';
import TableTree from '@atlaskit/table-tree';
import type { Item } from '../interfaces/interfaces';

const initialState = {
    uniqueIssues: jest.fn(),
    uniqueIssuesBy: jest.fn(),
    uniqueAssigneeBy: jest.fn(),
    uniqueLabelsBy: jest.fn(),
    uniquePriorityBy: jest.fn()
}

const shallowWrapper = shallow(<Statistics {...initialState}/>);

const stitisticsTableData = [
    {
        id: 'issues',
        content: ['Unique issues', 'issue', 5]
    },
    {
        id: 'assegnees',
        content: ['Unique assignees', 'assignee', userList.length]
    },
    {
        id: 'labels',
        content: ['Unique labels', 'labelIds', labels.length]
    },
    {
        id: 'priority',
        content: ['Unique priorities', 'priority', priorityList.length]
    }
];

describe('Statistics Table testing...', () => {

    it('Statistics Table should include TableTree (using "shallow")', () => {
        expect(shallowWrapper.find('TableTree')).toExist();
        expect(shallowWrapper.find('TableTree').length).toBe(1);
    })

    it('Statistics Table should include SNAPSHOT', () => {
        const tree = shallow(
        <TableTree
                headers={['Number of', 'Number by', 'Total number']}
                columns={['numberOf', 'numberBy', 'total']}
                columnWidths={['300px', '300px', '200px']}
                items={stitisticsTableData}
            />
            );
        
        expect(tree).toMatchSnapshot();
    })
})