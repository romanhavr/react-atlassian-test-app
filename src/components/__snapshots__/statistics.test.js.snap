// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics Table testing... Statistics Table should include SNAPSHOT 1`] = `
<styled__TableTreeContainer
  aria-readonly={true}
  role="treegrid"
>
  <Headers>
    <CellWithColumnWidth
      columnIndex={0}
      key="0"
      width="300px"
    >
      Number of
    </CellWithColumnWidth>
    <CellWithColumnWidth
      columnIndex={1}
      key="1"
      width="300px"
    >
      Number by
    </CellWithColumnWidth>
    <CellWithColumnWidth
      columnIndex={2}
      key="2"
      width="200px"
    >
      Total number
    </CellWithColumnWidth>
  </Headers>
  <Rows
    items={
      Array [
        Object {
          "content": Array [
            "Unique issues",
            "issue",
            5,
          ],
          "id": "issues",
        },
        Object {
          "content": Array [
            "Unique assignees",
            "assignee",
            4,
          ],
          "id": "assegnees",
        },
        Object {
          "content": Array [
            "Unique labels",
            "labelIds",
            3,
          ],
          "id": "labels",
        },
        Object {
          "content": Array [
            "Unique priorities",
            "priority",
            3,
          ],
          "id": "priority",
        },
      ]
    }
    render={[Function]}
  />
</styled__TableTreeContainer>
`;
